internal/controller/apiextensions/composite/composition_render.go

```diff
@@ -61,16 +61,18 @@ func RenderFromJSON(o resource.Object, data []byte) error {
 	o.SetName(name)
 	o.SetNamespace(namespace)
 
-	// This resource already had a GK (probably because it already exists), but
+	// This resource already had a Kind (probably because it already exists), but
 	// when we rendered its template it changed. This shouldn't happen. Either
-	// someone changed the kind or group in the template, or we're trying to use the
+	// someone changed the kind in the template, or we're trying to use the
 	// wrong template (e.g. because the order of an array of anonymous templates
 	// changed).
 	// Please note, we don't check for version changes, as versions can change. For example,
 	// if a composed resource was created with a template that has a version of "v1alpha1",
 	// and then the template is updated to "v1beta1", the composed resource will still be valid.
-	if !gvk.Empty() && o.GetObjectKind().GroupVersionKind().GroupKind() != gvk.GroupKind() {
-		return errors.Errorf(errFmtKindOrGroupChanged, gvk, o.GetObjectKind().GroupVersionKind())
+	// We also don't check for group changes, as groups can change during
+	// migrations.
+	if !gvk.Empty() && o.GetObjectKind().GroupVersionKind().Kind != gvk.Kind {
+		return errors.Errorf(errFmtKindChanged, gvk, o.GetObjectKind().GroupVersionKind())
 	}
```

Comment:
Changing the API version (e.g. v1beta1 -> v1) of a composed resource can be done in place. Crossplane will just start hitting the new v1 API endpoint to update the state.

You can't update the API group in place, as far as I know. So if you change the API group here Crossplane will:

* Create a new composed resource in the new API group
* Treat the old composed resource as no longer desired, and delete it

This enables migrating to a different API group, but the migration will be destructive. Are we comfortable with that?

Reply:
By enabling (group) migrations we also weaken a safety check.

Consider especially 'legacy' compositions where the composed templates don't have a `name`. Crossplane must then use the order of the `resources` array to figure out which template corresponds to which existing composed resource. If folks reorder that list (or remove an element from the middle) unexpected things would happen (e.g. composed resources would be deleted and recreated).

To be fair, I can see a case that just checking for `Kind` is good enough. It's unlikely that you accidentally switch two composed resources that are different but have the same `Kind`.

On the other hand, maybe it would be more explicit when doing a migration to a completely different kind of composed resource to change its composed resource name. This might make it more obvious that the existing composed resource will be deleted and replaced with a new one, not updated in place.

Reply:
We've running the migration process across several environments for a few months, and this process halted with Crossplane v1.14/v1.15 - We want the simplified method back to update composition to official provider resources without manually updating resourceRefs ( 120 resources * 25 Claims/XRs) or other migration steps. We have all official managed resources pre-created in an earlier step with same metadata.name, labels and annotations - the only step missing is now that the XR starts managing these pre-created official managed resources - which was working with v1.13. While you acknowledge the issues with real legacy compositions lacking `spec.resources[].name` - i understand this - but in real most of the environments using the `name` - and in this case, we're using the name to adopt the official managed resource

Reply:
I agree with the points @negz is making about weakening a safety check, on the other hand, I feel like that they are not that crucial in practice. My main argument for that is that we are running without such a check, up to the Crossplane version 1.13, and there are no known incidents related to that (AFAIU). 
So, I am leaning towards moving forward with this PR to facilitate the migration scenarios that @haarchri is pointing out.